### PR TITLE
Splits miner commits between stackers and a custodial address

### LINKED ISSUE #5: Split miner commits
- custodied wallet keys are only accessible to the respective city (mayor, treasurer, etc)
- if Stacking is active:
  - 70% of the miner commits are distributed to Stackers of the citycoin
  - 30% of the miner commits are distributed to the custodied wallet for the city ("THE BAG")
- if Stacking is not active:
  - 100% of the miner commits are distributed to the custodied wallet for the city ("THE BAG")

### PR DESCRIPTION
**Pull request by asteriabtc**
closes #5

adds two new constants:
- `STACKING_ACTIVE_STACKER_MULTIPLIER` - stacker proportion of miner commits, defaults to 0.7 (70%)
- `STACKING_ACTIVE_BAG_MULTIPLIER` - bag/custodial address proportion of miner commits, defaults to 0.3 (30%)

adds new `read-only` method:
- `stacking-is-active` - returns true if stacking is active, false if not

modifies the `mine-tokens` method to split miner commits between the stackers and bag holder in ratios defined by `STACKING_ACTIVE_STACKER_MULTIPLIER` & `STACKING_ACTIVE_BAG_MULTIPLIER` constants.

### CODE REVIEW

**Comment by whoabuddy** [comment]
Awesome will take a look at this tomorrow!

**Comment by LNow** [comment]
I see one flaw in this implementation.
If for some reason there will be no stackers in a given cycle, then 70% of the funds paid by miners in this cycle will get locked in contract... forever. I think that in such case 100% should go to the city.

IMHO all STX committed by miners should be saved inside contract like in original implementation and splitting should be done either virtually (by saving in map or variables how much should go to bag holder and how much to stackers) or while making payouts to stackers (by calculating reward based on 70% of all STX commited by miners)
And we should add new public function that will allow bag holder (the City) to withdraw all funds not assigned for stacking rewards.

**Comment by pxydn** [comment]
so (1) we need a minimum amount of stackers, maybe 32, the same as the minimum miners. shouldn't be a lot of code to change. i'll do this. (2) you think we shouldn't split miner funds/stacking rewards directly, rather keep them in the contract till they are redeemed. what do you think of this @whoabuddy? personally i prefer the current implementation imo

File: citycoin.clar [hunk 1]
```diff
@@ -22,6 +22,9 @@
 (define-constant FIRST-STACKING-BLOCK u1)           ;; Stacks block height when Stacking is available
 (define-constant REWARD-CYCLE-LENGTH u500)          ;; how long a reward cycle is
 (define-constant MAX-REWARD-CYCLES u32)             ;; how many reward cycles a Stacker can Stack their tokens for
+(define-constant THE_BAG principal ')               ;; the custodied wallet for the city
+(define-constant STACKING_ACTIVE_STACKER_MULTIPLIER 0.7)
+(define-constant STACKING_ACTIVE_BAG_MULTIPLIER 0.3)
```
Comment (jcnelson) [inline citycoin.clar:27]: Clarity does not have a floating-point type.

File: citycoin.clar [hunk 2]
```diff
@@ -569,7 +575,15 @@
         (try! (can-mine-tokens tx-sender block-height amount-ustx miner-rec))
 
         (try! (set-tokens-mined tx-sender block-height amount-ustx))
-        (unwrap-panic (stx-transfer? amount-ustx tx-sender (as-contract tx-sender)))
+
+        (if (stacking-is-active)
+            (begin 
+                (unwrap-panic (stx-transfer? STACKING_ACTIVE_STACKER_MULTIPLIER tx-sender (as-contract tx-sender)))
+                (unwrap-panic (stx-transfer? STACKING_ACTIVE_BAG_MULTIPLIER tx-sender THE_BAG))
```
Comment (jcnelson) [inline citycoin.clar:582]: Forget to multiply? ;)

**Comment by whoabuddy** [comment]
If you can resolve conflicts we can get this one merged as well!

**Comment by whoabuddy** [comment]
Closing in favor of #45 since a lot has changed since this PR started 😬

**Pull request closed without merging**